lib/irb.rb

```diff
@@ -485,6 +485,11 @@ def run(conf = IRB.conf)
       conf[:IRB_RC].call(context) if conf[:IRB_RC]
       conf[:MAIN_CONTEXT] = context
 
+      if conf[:SAVE_HISTORY] && context.io.support_history_saving?
+        context.io.load_history
+        conf[:AT_EXIT].push proc{ context.io.save_history }
```

Comment:
Note: `conf[:AT_EXIT]` will be called when `Irb#run` exits and is not associated with Ruby's `at_exit` callback.